### PR TITLE
WC2-119 Pluralized all APIs which are not tasks or functions

### PR DESCRIPTION
**Pull request by mathvdh**
All API endpoints should have names in plural form !
This is a bit dangerous, especially for endpoints used by mobile applications.

Related JIRA tickets : WC2-119
## Self proof reading checklist

- [x] Did I use eslint and black formatters
- [x] Are there enough tests

## Changes

The following endpoints were changed, along with frontend and eventual tests 

- [x] router.register(r"devicesownership" ..) 
- [x] router.register(r"devicesposition" ..)
- [x] router.register(r"entity" ..)
- [x] router.register(r"entitytype" ..)
- [x] router.register(r"bulkcreateuser" ..) 
- [x] router.register(r"storage" ..)
- [x] router.register(r"microplanning/planning", PlanningViewSet, basename="planning")
- [x] path("storages/<str:storage_type>/<str:storage_customer_chosen_id>/logs", logs_per_device)

For the mobiles endpoints plural aliases were created and swagger doc was updated with a deprecation notice

- [x] router.register(r"mobile/entitytype" ..)
- [x] router.register(r"mobile/checkversion" ..)
- [x] router.register(r"mobile/storage/logs" ..)
- [x] router.register(r"mobile/storage/blacklisted" ..)

The following endpoints where **NOT** changed (tasks or functions)

~~router.register(r"copyversion" ..)~~
~~router.register(r"setupaccount" ..)~~
~~router.register(r"dhis2ouimporter" ..)~~
~~router.register(r"tasks/create/orgunitsbulkupdate" ..)~~
~~router.register(r"tasks/create/importgpkg" ..)~~

## How to test

Well almost every functionality that uses any of these endpoints should be tested manually I guess (python tests and cypress tests were run)

### CODE REVIEW

**Comment by mathvdh** [comment]
@cypress

**Comment by mathvdh** [comment]
Cypress Hill

**Review by beygorghor — APPROVED** [review]

File: iaso/api/devices_position.py [hunk 1]
```diff
@@ -55,8 +55,8 @@ class DevicesPositionViewSet(ModelViewSet):
 
     This API is open to anonymous users on write (if the project allows it).
 
-    GET /api/devicesposition/
-    POST /api/devicesposition/
+    GET /api/devicespositions/
+    POST /api/devicespositions/
```
Comment (bmonjoie) [inline iaso/api/devices_position.py:61]: The mobile application uses this one. It should still be callable without the `s` (at least for the `POST`)

File: iaso/api/storage.py [hunk 2]
```diff
@@ -349,7 +349,8 @@ class StorageLogViewSet(CreateModelMixin, viewsets.GenericViewSet):
 
     def create(self, request):
         """
-        POST /api/mobile/storage/logs
+        POST /api/mobile/storage/logs [Deprecated] will be removed in the future
```
Comment (bmonjoie) [inline iaso/api/storage.py:351]: Nitpicking here but everywhere else you put the deprecated endpoint under the new one and not above like here

File: iaso/api/check_version.py [hunk 3]
```diff
@@ -18,7 +18,10 @@ class CheckVersionViewSet(ViewSet):
 
     This API is open to anonymous users:
 
-    `GET /api/mobile/checkversion/?app_id=some.app.id&app_version=1234`
+    /api/mobile/checkversions/
+    /api/mobile/checkversion/ [Deprecated] will be removed in the future
+
+    `GET /api/mobile/checkversions/?app_id=some.app.id&app_version=1234`
```
Comment (bmonjoie) [inline iaso/api/check_version.py]: Here, it's not a list of resources but a check on the version. It should not be pluralized.

File: iaso/urls.py [hunk 4]
```diff
@@ -113,28 +113,28 @@
...
+router.register(r"entities", EntityViewSet, basename="entity")
 router.register(r"mobile/entities", MobileEntityViewSet, basename="entities")
-router.register(r"entitytype", EntityTypeViewSet, basename="entitytype")
+router.register(r"entitytypes", EntityTypeViewSet, basename="entitytype")
 # At the moment we use the same view set but separate it for the future for when we want to be able to
 # change the format in the future
-router.register(r"mobile/entitytype", EntityTypeViewSet, basename="entitytype")
-router.register(r"bulkcreateuser", BulkCreateUserFromCsvViewSet, basename="bulkcreateuser")
+router.register(r"mobile/entitytypes?", EntityTypeViewSet, basename="entitytype")
+router.register(r"bulkcreateusers", BulkCreateUserFromCsvViewSet, basename="bulkcreateuser")
```
Comment (bmonjoie) [inline iaso/urls.py]: Same remark as for checkversion, It's not a list of resources but an action, I would not pluralize

**Review by bmonjoie — CHANGES_REQUESTED** [review]
I might be wrong but I see you removed the routes for the mobile endpoints from iaso/urls.py that should still be singular. Therefore, won't this make them unavailable and break?

**Comment by mathvdh** [comment]
@bmonjoie thanks for the input, I will fix tomorrow. Regarding `/api/devicesposition/` Shouldn't it be moved to `/api/mobile/devicesposition/` for consistency ? I mean all the other mobile endpoints follow this pattern no ? Or maybe there is a specific reason .. ?

**Comment by olethanh** [comment]
@bmonjoie Could you please give a list of endpoint used by the mobile app??? (all active versions of the app)

**Comment by bmonjoie** [comment]
@olethanh 

v1.56
=====
POST /api/mobile/orgunits/
POST /api/instances/
GET /api/mobile/orgunits/
GET /api/forms/
GET /api/orgunittypes/
POST /api/devicesposition/
POST /api/token/
GET /api/apps/{app_id}/

2.0
===
POST /api/mobile/orgunits/
POST /api/instances/
GET /api/mobile/orgunits/
GET /api/mobile/forms/
GET /api/orgunittypes/
POST /api/devicesposition/
POST /api/token/
GET /api/apps/{app_id}/
GET /api/apps/current/

2.1
====
POST /api/mobile/orgunits/
POST /api/instances/
GET /api/mobile/orgunits/
GET /api/mobile/forms/
GET /api/orgunittypes/
POST /api/devicesposition/
POST /api/token/
GET /api/apps/{app_id}/
GET /api/apps/current/
GET /api/mobile/plannings/


In development
============
POST /api/mobile/orgunits/
POST /api/instances/
GET /api/mobile/orgunits/
GET /api/mobile/forms/
GET /api/formversions/
GET /api/orgunittypes/
POST /api/devicesposition/
POST /api/token/
GET /api/apps/{app_id}/
GET /api/apps/current/
GET /api/mobile/plannings/
GET /api/mobile/checkversion/
GET /api/mobile/entitytype/
GET /api/mobile/entities/
GET /api/mobile/storage/blacklisted/
POST /api/mobile/storage/logs/
GET /api/mobile/workflows/
GET /api/profiles/me/
GET /api/mobile/reports/

Those only in development can change but please let me know first

**Comment by olethanh** [comment]
I put in in the form of a table here. I find that a bit more readable

https://docs.google.com/spreadsheets/d/124okkt41o1bs6x_pvzI8QKB18kdsfBRUrh6h81Ro_Iw/edit#gid=0

We should have a test for each one to ensure we don't remove, change or rename them by error.

Also we should aim to have the mobile app use only the endpoint starting with `/api/mobile/`. especially for the one still only used in dev so we don't have to pay the cost twice


image

**Comment by mathvdh** [comment]
@bmonjoie  I made all the changes requested, I think .. So can you re-check it so the PR can eventually be merged ?

File: iaso/tests/api/test_check_version.py [hunk 5]
```diff
@@ -16,79 +19,79 @@ def setUpTestData(cls):
         )
 
     def test_check_version_delete(self):
-        """DELETE /api/mobile/checkversion should result in a 405 response"""
+        """DELETE /api/mobile/checkversions should result in a 405 response"""
```
Comment (bmonjoie) [inline iaso/tests/api/test_check_version.py]: Why is this still plural?

This applies to the whole file.

Comment (olethanh) [reply iaso/tests/api/test_check_version.py]: corrected

**Pull request merged**